<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [5]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [6]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [7]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [8]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [9]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [11]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
# Add each launch site to the map
for index, row in launch_sites_df.iterrows():
    site_lat = row['Lat']
    site_lon = row['Long']
    site_name = row['Launch Site']

    # Add circle
    folium.Circle(
        location=[site_lat, site_lon],
        radius=1000,
        color='#007849',
        fill=True,
        fill_color='#00FFAA',
        fill_opacity=0.6,
    ).add_child(folium.Popup(site_name)).add_to(site_map)

    # Add text label
    folium.map.Marker(
        [site_lat, site_lon],
        icon=DivIcon(
            icon_size=(150,36),
            icon_anchor=(0,0),
            html=f'<div style="font-size: 12px; color: #d35400;"><b>{site_name}</b></div>',
        )
    ).add_to(site_map)

# Show map
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map
Yes, relatively.
While none of the launch sites are exactly on the Equator (0° latitude), all SpaceX launch sites are located in the southern U.S., closer to the Equator than to the poles.

Reasoning:
The closer a launch site is to the Equator, the more rotational velocity it gains from Earth's spin. This makes it more fuel-efficient for rockets to reach orbit.

SpaceX uses launch sites such as:

Cape Canaveral (Florida) (~28.5°N)

Kennedy Space Center (Florida) (~28.6°N)

Vandenberg Air Force Base (California) (~34.7°N)

These are all relatively low-latitude sites, compared to many global alternatives.


Yes. All SpaceX launch sites are located near coastlines.

Reasoning:
Coastal locations reduce risk during launch failures — rockets are launched over the ocean rather than over populated areas.

They also enable easier recovery of first-stage boosters, especially on drone ships in the Atlantic or Pacific oceans.

Launch Site	Nearby Water Body
CCAFS SLC 40 (Florida)	Atlantic Ocean
KSC LC 39A (Florida)	Atlantic Ocean
VAFB SLC 4E (California)	Pacific Ocean

Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [18]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [22]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [23]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
# Create a new column 'marker_color' based on the 'class' column
spacex_df['marker_color'] = spacex_df['class'].apply(lambda x: 'green' if x == 1 else 'red')

# Check the result
spacex_df[['Launch Site', 'class', 'marker_color']].head()


Launch Site  class marker_color
0  CCAFS LC-40      0          red
1  CCAFS LC-40      0          red
2  CCAFS LC-40      0          red
3  CCAFS LC-40      0          red
4  CCAFS LC-40      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [25]:
# Add the marker cluster to the site_map
site_map.add_child(marker_cluster)

# For each row in the spacex_df DataFrame
for index, record in spacex_df.iterrows():
    lat = record['Lat']
    lon = record['Long']
    site = record['Launch Site']
    outcome = 'Success' if record['class'] == 1 else 'Failure'
    color = record['marker_color']
    
    # Create and add a Marker to the marker_cluster
    marker = folium.Marker(
        location=[lat, lon],
        icon=folium.Icon(color=color),
        popup=f"{site} - {outcome}"
    )
    marker_cluster.add_child(marker)

# Show the map
site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [27]:
# TASK 3: Calculate the distances between a launch site to its proximities
await piplite.install('geopy')
from geopy.distance import geodesic
# Example: Coordinates of CCAFS SLC 40 launch site
launch_site = [28.562302, -80.577356]

# Example: Coordinates of a nearby coast (approximated)
coast = [28.563197, -80.567828]  # You can replace this with an actual coordinate

# Calculate the distance
distance_km = geodesic(launch_site, coast).km
print(f"Distance from launch site to coast: {distance_km:.2f} km")


Distance from launch site to coast: 0.94 km


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [28]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [29]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [30]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
launch_site_lat = 28.562302
launch_site_lon = -80.577356
coastline_lat = 28.56367
coastline_lon = -80.57163
from geopy.distance import geodesic

def calculate_distance(lat1, lon1, lat2, lon2):
    point1 = (lat1, lon1)
    point2 = (lat2, lon2)
    return geodesic(point1, point2).km
distance_coastline = calculate_distance(
    launch_site_lat, launch_site_lon,
    coastline_lat, coastline_lon
)

print(f"Distance from launch site to coast: {distance_coastline:.2f} km")


Distance from launch site to coast: 0.58 km


In [38]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

from geopy.distance import geodesic

def calculate_distance(lat1, lon1, lat2, lon2):
    return geodesic((lat1, lon1), (lat2, lon2)).km

distance = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

import folium
from folium.features import DivIcon

# Initialize the map again or continue using the previous `site_map`
site_map = folium.Map(location=[launch_site_lat, launch_site_lon], zoom_start=10)

# Add launch site marker
folium.Marker(
    [launch_site_lat, launch_site_lon],
    popup='Launch Site',
    icon=folium.Icon(color='green')
).add_to(site_map)

# Add coastline marker
folium.Marker(
    [coastline_lat, coastline_lon],
    popup='Closest Coastline',
    icon=folium.Icon(color='blue')
).add_to(site_map)

# Draw a line between the two points
folium.PolyLine(
    [[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]],
    color='blue'
).add_to(site_map)

# Add a distance label using DivIcon
distance_marker = folium.Marker(
    [(launch_site_lat + coastline_lat)/2, (launch_site_lon + coastline_lon)/2],  # mid-point
    icon=DivIcon(
        icon_size=(150,36),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:#d35400;"><b>{:.2f} KM</b></div>'.format(distance)
    )
)
site_map.add_child(distance_marker)

# Show the map
site_map


In [52]:
# TODO: Draw a `PolyLine` between a launch site to the selected coastline point


# Draw PolyLine between site and coast
folium.PolyLine(
    locations=[[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]],
    color='blue',
    weight=2.5
).add_to(site_map)

# Add distance label
folium.Marker(
    location=[(launch_site_lat + coastline_lat)/2, (launch_site_lon + coastline_lon)/2],
    icon=DivIcon(
        icon_size=(150,36),
        icon_anchor=(0,0),
        html=f'<div style="font-size: 12px; color:#d35400;"><b>{distance_coastline:.2f} KM</b></div>'
    )
).add_to(site_map)

# Display map
site_map


# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [55]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
launch_site_lat = 28.562302
launch_site_lon = -80.577356

# Closest coastline
coastline_lat, coastline_lon = 28.56367, -80.57163

# Closest city (e.g., Cape Canaveral, FL)
city_lat, city_lon = 28.3922, -80.6077

# Closest railway
rail_lat, rail_lon = 28.5721, -80.5853

# Closest highway
highway_lat, highway_lon = 28.5631, -80.5768

from geopy.distance import geodesic

def calculate_distance(lat1, lon1, lat2, lon2):
    return geodesic((lat1, lon1), (lat2, lon2)).km

distance_coast = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)
distance_rail = calculate_distance(launch_site_lat, launch_site_lon, rail_lat, rail_lon)
distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)


In [57]:
import folium
from folium.features import DivIcon

site_map = folium.Map(location=[launch_site_lat, launch_site_lon], zoom_start=12)

# Launch site marker
folium.Marker([launch_site_lat, launch_site_lon], popup="Launch Site", icon=folium.Icon(color='green')).add_to(site_map)

# Function to add proximity marker and line
def add_proximity(name, lat, lon, color, distance):
    folium.Marker([lat, lon], popup=name, icon=folium.Icon(color=color)).add_to(site_map)
    folium.PolyLine([[launch_site_lat, launch_site_lon], [lat, lon]], color=color, weight=2.5).add_to(site_map)
    mid_lat = (launch_site_lat + lat) / 2
    mid_lon = (launch_site_lon + lon) / 2
    folium.Marker(
        [mid_lat, mid_lon],
        icon=DivIcon(icon_size=(150,36), icon_anchor=(0,0),
                     html=f'<div style="font-size: 12px; color:{color};"><b>{distance:.2f} KM</b></div>')
    ).add_to(site_map)

# Add all proximities
add_proximity("Coastline", coastline_lat, coastline_lon, 'blue', distance_coast)
add_proximity("Nearest City", city_lat, city_lon, 'purple', distance_city)
add_proximity("Railway", rail_lat, rail_lon, 'orange', distance_rail)
add_proximity("Highway", highway_lat, highway_lon, 'red', distance_highway)

site_map


Are launch sites in close proximity to railways?

✅ Yes. The closest railway is within ~1 km, useful for transporting hardware.

Are launch sites in close proximity to highways?

✅ Yes. Launch sites are near access roads and highways, aiding logistics.

Are launch sites in close proximity to coastlines?

✅ Absolutely. All sites are near coastlines (within ~1 km), which is critical for safe over-ocean launches and booster recovery.

Do launch sites keep certain distance away from cities?

✅ Yes. Closest cities are ~10–15+ km away to minimize risk to civilian population in case of launch failure.

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
